src/Xrefcheck/Verify.hs

```diff
@@ -255,32 +284,87 @@ verifyReference
-
-    let moveProgress = incProgress .
-          (if verifyOk res then id else incProgressErrors)
-
-    atomicModifyIORef' progressRef $ \VerifyProgress{..} ->
-      ( if isExternal locType
-        then VerifyProgress{ vrExternal = moveProgress vrExternal, .. }
-        else VerifyProgress{ vrLocal = moveProgress vrLocal, .. }
-      , ()
-      )
-    return $ fmap (WithReferenceLoc fileWithReference ref) res
-  else return mempty
+    = retryVerification 0 $ do
+        let locType = locationType rLink
+        if shouldCheckLocType mode locType
+        then case locType of
+          LocalLoc    -> checkRef rAnchor fileWithReference
+          RelativeLoc -> checkRef rAnchor
+                        (takeDirectory fileWithReference
+                          </> toString (canonizeLocalRef rLink))
+          AbsoluteLoc -> checkRef rAnchor (root <> toString rLink)
+          ExternalLoc -> checkExternalResource config rLink
+          OtherLoc    -> verifying pass
+        else return mempty
   where
+    retryVerification
+      :: Int
+      -> IO (VerifyResult VerifyError)
+      -> IO (VerifyResult $ WithReferenceLoc VerifyError)
+    retryVerification r resIO = do
+      res@(VerifyResult ves) <- resIO
+
+      let toRetry = any isFixable ves && r < maxRetries
+          currentDelay = extractRetryAfterInfo res
+
+      let moveProgress = alterOverallProgress r
+                       . alterProgressErrors res r
+
+      posixTime' <- getPOSIXTime
+      let posixTime = posixTimeToTimeSecond posixTime'
+
+      atomicModifyIORef' progressRef $ \VerifyProgress{..} ->
+        ( if isExternal $ locationType rLink
+          then VerifyProgress{ vrExternal =
+            let vrExternalAdvanced = moveProgress vrExternal
+            in if toRetry
+               then let mTTC = fmap fst . pTaskTimestamp $ vrExternal
+                        maxTime = maybe currentDelay (currentDelay `max`) mTTC
+                    in if maxTime == currentDelay
+                       then setTaskTimestamp maxTime posixTime vrExternalAdvanced
+                       else vrExternalAdvanced
+               else vrExternalAdvanced, .. }
+          else VerifyProgress{ vrLocal = moveProgress vrLocal, .. }
+        , ()
+        )
+      if toRetry
+      then do
+        threadDelay currentDelay
+        retryVerification (r + 1) resIO
+      else return $ fmap (WithReferenceLoc fileWithReference ref) res
+
+    alterOverallProgress
+      :: (Num a, Show a)
+      => Int
+      -> Progress a
+      -> Progress a
+    alterOverallProgress retryNumber
+      | retryNumber > 0 = id
+      | otherwise = incProgress
+
+    alterProgressErrors
+      :: (Num a, Show a)
+      => VerifyResult VerifyError
+      -> Int
+      -> Progress a
+      -> Progress a
+    alterProgressErrors res@(VerifyResult ves) retryNumber
+      | retryNumber >= maxRetries && any isFixable ves ||
+        retryNumber > 0 && any isUnfixable ves = fixableToUnfixable
```

Comment:
I'm not sure if that's what you meant to do but, if you meant to say the opposite `any isFixable ves`, then that should be "**all** isUnfixable ves"

Reply:
Oh, I meant `any isUnfixable ves` because there's a base case `all _ [] = True`, and if there are no errors, instead of decreasing the number of fixable errors, we would then be transforming it to unfixable because of the guards. Since `verifyReference` returns either a singleton list or an empty list, wrapped in `VerifyResult`, `any isUnfixable ves` will only return `True` if `ves` is a singleton list and contains an unfixable error.
But nonetheless, as you've mentioned in your comment above this one, the whole guard business should be refactored top to bottom, which is what I'm going to do.